udata/api_fields.py

```diff
@@ -0,0 +1,247 @@
+        cls.__write_fields__ = api.model(f"{cls.__name__} (write)", write_fields, **kwargs)
+
+        mask = kwargs.pop('mask', None)
+        if mask is not None:
+            mask = 'data{{{0}}},*'.format(mask)
+        cls.__page_fields__ = api.model(f"{cls.__name__}Page", custom_restx_fields.pager(cls.__read_fields__), mask=mask, **kwargs)
+
+        # Parser for index sort/filters
+        paginable = kwargs.get('paginable', True)
+        parser = api.parser()
+
+        if paginable:
+            parser.add_argument('page', type=int, location='args', default=1, help='The page to display')
+            parser.add_argument('page_size', type=int, location='args', default=20, help='The page size')
+        
+        if sortables:
+            choices = sortables + ['-' + k for k in sortables]
+            parser.add_argument('sort', type=str, location='args', choices=choices, help='The field (and direction) on which sorting apply')
+
+        cls.__index_parser__ = parser
+        def apply_sort_filters_and_pagination(base_query):
+            args = cls.__index_parser__.parse_args()
+
+            if sortables and args['sort']:
+                base_query = base_query.order_by(args['sort'])
+
+            if paginable:
+                base_query = base_query.paginate(args['page'], args['page_size'])
+
+            return base_query
+
+        cls.apply_sort_filters_and_pagination = apply_sort_filters_and_pagination
+        return cls
+    return wrapper
+
+
+def function_field(**info):
+    def inner(func):
+        func.__additional_field_info__ = info
+        return func
+
+    return inner
+
+def field(inner, **kwargs):
+    '''
+    Simple decorator to mark a field as visible for the API fields.
+    We can pass additional arguments that will be forward to the RestX field constructor.
+    '''
+    inner.__additional_field_info__ = kwargs
+    return inner
+
+
+def patch(obj, request): 
+    '''
+    Patch the object with the data from the request.
+    Only fields decorated with the `field()` decorator will be read (and not readonly).
+    '''
+    for key, value in request.json.items():
+        field = obj.__write_fields__.get(key)
+        if field is not None and not field.readonly:
+            model_attribute = getattr(obj.__class__, key)
+            if isinstance(model_attribute, mongoengine.fields.ListField) and isinstance(model_attribute.field, mongoengine.fields.ReferenceField):
+                # TODO `wrap_primary_key` do Mongo request, do a first pass to fetch all documents before calling it (to avoid multiple queries).
+                value = [wrap_primary_key(key, model_attribute.field, id) for id in value]
+            if isinstance(model_attribute, mongoengine.fields.ReferenceField):
+                value = wrap_primary_key(key, model_attribute, value)
+
+
+            info = getattr(model_attribute, '__additional_field_info__', {})
+            check = info.get('check', None)
+            if check is not None:
```

Comment:
Unsure when check is set?

Reply:
`check` allows to do validation in `Owned.py` for organization and owner

Reply:
Added a little comment in 335b309c4